### ISSUE TITLE
Why create an element with a unique id, which is already available?

### ISSUE THREAD

**Issue by Riim** [patrick-steele-idem/morphdom#30]
```js
<!DOCTYPE html>
<html>
<body>

<script src="node_modules/morphdom/dist/morphdom-umd.js"></script>
<script>

var fromEl = document.createElement('div');
var toEl = document.createElement('div');

fromEl.innerHTML = '<div id="qwerty" class="div1"></div>';
toEl.innerHTML = '<div><div id="qwerty" class="div1"></div></div>';

var div1 = fromEl.querySelector('.div1');

morphdom(fromEl, toEl);

var div1_2 = fromEl.querySelector('.div1');

console.log(div1 == div1_2); // false ???, should be true

</script>

</body>
</html>
```

**Comment by patrick-steele-idem** [comment]
Hmm. Seems like a bug if that is the case. I'm looking into it.

**Comment by patrick-steele-idem** [comment]
Hey @Riim, I've done some preliminary investigation and it does look like there is a gap in the code. Ideally, we should match up all elements with matching keys/IDs and this is not working in some cases. I added some failing test cases and am working on a fix. Thank you for reporting the problem and providing test code to easily reproduce!

**Comment by Riim** [comment]
As far as I was able to understand the algorithm, it has the probability of such errors.

I rewrote in its own way can be useful to you my version: https://github.com/Riim/rista/blob/master/src/morphElement.js

This version guarantees the use of already available elements with a unique key. But unfortunately about 30 percent slower. At the same time I didn't do whatever or optimization and can be easily accelerated.

**Comment by patrick-steele-idem** [comment]
Hey @Riim, I pushed some changes to `morphdom` to ensure that the keyed elements are matched up correctly in [hopefully] all cases. Please check out the new code that has been merged. I still need to do some more testing in a real app before I publish a new version.

As you probably noticed, the morphdom module uses some heuristics to be fastest in the "normal" case which is what makes the code more complicated than a straightforward algorithm. The most important thing for morphdom is that the DOM matches up at the end and at least that was working correctly before. If we can match up all keyed elements then that is ideal, but it definitely makes the code more tricky if you try to do it in very performant way. Hopefully it works now.

Also, I took a look at your code and it has some good ideas that we may be able to bring in to morphdom, but we definitely don't want to introduce code that makes it 30% slower (of course, it likely varies on the DOM trees being diffed/transformed).

Let me know if you find any problems or have suggestions on how to improve the code. Thank you for your contributions!

**Comment by patrick-steele-idem** [comment]
I was slow getting a new version of this module published with the fix because of the holidays, but I just published `morphdom@1.0.4`. The new version fixes the problem that you reported. It also turned out that this issue was also impacting Marko Widgets so thank you for reporting the problem! I also added a much needed CHANGELOG: https://github.com/patrick-steele-idem/morphdom/blob/master/CHANGELOG.md
